hail/src/main/scala/is/hail/expr/ir/TableIR.scala

```diff
@@ -851,7 +852,71 @@ case class TableMapRows(child: TableIR, newRow: IR) extends TableIR {
         scanInitOps(0)(region, scanAggs, globals, false)
       }
 
-      val itF = { (i: Int, ctx: RVDContext, partitionAggs: Array[RegionValueAggregator], it: Iterator[RegionValue]) =>
+      val locallyScannedFile = HailContext.get.getTemporaryFile()
+
+      val (partFiles, _) = tv.rvd.crdd.cmapPartitionsWithIndex { (i, ctx, it) =>
+        val globals =
+          if (scanSeqNeedsGlobals) {
+            val rvb = new RegionValueBuilder(ctx.freshRegion)
+            rvb.start(gType.physicalType)
+            rvb.addAnnotation(gType, globalsBc.value)
+            rvb.end()
+          } else
+              0
+
+        val scanSeqOpF = scanSeqOps(i)
+        it.foreach { rv =>
+          scanSeqOpF(rv.region, scanAggs, globals, false, rv.offset, false)
+          ctx.region.clear()
+        }
+        Iterator.single(scanAggs)
+      }.writePartitions(locallyScannedFile, true, { (ctx, it, os) =>
+        val aggs = it.next
+        new ObjectOutputStream(os).writeObject(aggs)
+        1
+      })
+
+      val hConf = hc.hadoopConf
+      val inputStreams = new Array[ObjectInputStream](partFiles.length)
+      val globallyScannedFile = HailContext.get.getTemporaryFile()
+      try {
+        hConf.writeFile(globallyScannedFile) { os =>
+          using(new ObjectOutputStream(os)) { oos =>
```

Comment:
It seems like you can get rid of `inputStreams` and open one file at a time if you do, in pseudo-code:

```
  partFiles.indices.scanLeft(scanAggs) { (a, i) => 
    val a2 = using(open part file) { ois => load ith agg from part file }
    newA = a.copy()
    newA.combOp(a2)
    newA
}
```

or something like that.

Man, that copy is unfortunate.

Reply:
ah good point